<html>
  <head>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="https://fonts.googleapis.com/css2?family=Jost&display=swap"
          rel="stylesheet">
    <link href="https://fonts.googleapis.com/css2?family=Fira+Mono:wght@400;500&family=Nova+Mono&family=Share+Tech+Mono&display=swap"
          rel="stylesheet">
    <link rel="stylesheet"
          href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/10.0.2/styles/a11y-light.min.css">
    <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/10.0.2/highlight.min.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <link href="/style.css" rel="stylesheet">
  </head>
  <body>
    <div class="title"><a href="/"><h1>Hail</h1></a></div>

    <div class="background"><div class="content">
        <h3>Introduction for Scientists</h3>

        <p>
          Hail is a Python library for the interactive analysis of genotype,
          whole-exome-sequencing, and whole-gneome-sequencing datasets of any
          size. Hail can import many genetics formats including VCF, UKB BGEN,
          and PLINK. Hail enables variant and sample selection based on
          arbitrary sample, variant, allele, or genotype statistics. Hail
          supports linear, logistic and poisson regression.
        </p>
        <p>
          Hail annotates. The Hail team maintains many annotation datasets in
          native Hail format for low-effort annotation onto your genetic
          datasets. One line of Hail can load ten different genetic annotations.
          Analysts are freed from finding, harmonize, or importing data. The
          data is simply annotated where it belongs.
        </p>
        <p>
          Hail scales. Hail can leverage clusters of a thousand machines to
          analyze data quickly. Fast computation allows rapid iteration on QC
          and regression parameters leading to higher quality results.
        </p>
        <p>
          Hail adapts. In addition to the standard VCF-like datatypes, Hail can
          also work with gene-sample, gene-population, gene-phenotype, and
          variant-phenotype datasets.
        </p>
        <p>
          Hail matrix multiplies. Hail enables the calculation of huge matrix
          multiplies, such as a banded LD Matrix of tens of millions of variants
          or a pruned relationship matrix of hundreds of thousands of samples.
        </p>
        <ul class="button-list">
          <li><a href="install.html">Install</a></li>
          <li><a href="get-help.html">Get Help</a></li>
        </ul>
        <p>
          These capabilities and more has made Hail the foundation of many
          scientific papers. Some of them are listed here.
        </p>
        <h2 id="section">2020</h2>
        <ul>
          <li>Muniz Carvalho, C., Wendt, F.R., Maihofer, A.X. et al. Dissecting the genetic association of C-reactive protein with PTSD, traumatic events, and social support. <em>Neuropsychopharmacol.</em> (2020). <a href="https://doi.org/10.1038/s41386-020-0655-6" class="uri">https://doi.org/10.1038/s41386-020-0655-6</a> <a href="https://www.nature.com/articles/s41386-020-0655-6#citeas" class="uri">https://www.nature.com/articles/s41386-020-0655-6#citeas</a>
          </li>
        </ul>
        <h2 id="section-1">2019</h2>
        <ul>
          <li><p>Gay, Nicole R. et al. "Impact of admixture and ancestry on eQTL analysis and GWAS colocalization in GTEx" <em>bioRxiv</em> (2019) 836825; <a href="https://www.biorxiv.org/content/10.1101/836825v1" class="uri">https://www.biorxiv.org/content/10.1101/836825v1</a></p></li>
          <li><p>Sakaue, Saori et al. "Trans-biobank analysis with 676,000 individuals elucidates the association of polygenic risk scores of complex traits with human lifespan" <em>bioRxiv</em> (2019): 856351 <a href="https://www.biorxiv.org/content/10.1101/856351v1" class="uri">https://www.biorxiv.org/content/10.1101/856351v1</a></p></li>
          <li><p>Polimanti, Renato et al. "Leveraging genome-wide data to investigate differences between opioid use vs. opioid dependence in 41,176 individuals from the Psychiatric Genomics Consortium" <em>bioRxiv</em> (2019): 765065 <a href="https://www.biorxiv.org/content/10.1101/765065v1" class="uri">https://www.biorxiv.org/content/10.1101/765065v1</a></p></li>
          <li><p>Wang, Qingbo et al. "Landscape of multi-nucleotide variants in 125,748 human exomes and 15,708 genomes" <em>bioRxiv</em> (2019): 573378; https://www.biorxiv.org/content/10.1101/573378v2.full</p></li>
          <li><p>Lescai, Francesco et al. "Meta-analysis of Scandinavian Schizophrenia Exomes" <em>bioRxiv</em> (2019): 836957; <a href="https://www.biorxiv.org/content/10.1101/836957v2" class="uri">https://www.biorxiv.org/content/10.1101/836957v2</a></p></li>
          <li><p>Bolze, Alexandre, et al. "Selective constraints and pathogenicity of mitochondrial DNA variants inferred from a novel database of 196,554 unrelated individuals" <em>bioRxiv</em> (2019): 798264;<a href="https://www.biorxiv.org/content/10.1101/798264v1" class="uri">https://www.biorxiv.org/content/10.1101/798264v1</a></p></li>
          <li><p>De Lillo, A., De Angelis, F., Di Girolamo, M. et al. "Phenome-wide association study of TTR and RBP4 genes in 361,194 individuals reveals novel insights in the genetics of hereditary and wildtype transthyretin amyloidoses." <em>Hum Genet</em> 138, 1331–1340 (2019). <a href="https://www.ncbi.nlm.nih.gov/pubmed/31659433" class="uri">https://www.ncbi.nlm.nih.gov/pubmed/31659433</a></p></li>
          <li><p>Pividori, Milton, et al. "Shared and distinct genetic risk factors for childhood-onset and adult-onset asthma: genome-wide and transcriptome-wide studies." <em>The Lancet Respiratory Medicine</em> 7.6 (2019): 509-522. <a href="https://www.biorxiv.org/content/10.1101/427427v2" class="uri">https://www.biorxiv.org/content/10.1101/427427v2</a></p></li>
          <li><p>Werling, Donna, et al. "Whole-genome and RNA sequencing reveal variation and transcriptomic coordination in the developing human prefrontal cortex." <em>bioRxiv</em> (2019): 538421. <a href="https://www.biorxiv.org/content/10.1101/585430v1" class="uri">https://www.biorxiv.org/content/10.1101/585430v1</a></p></li>
          <li><p>Satterstrom, Kyle F., et al. "Large-scale exome sequencing study implicates both developmental and functional changes in the neurobiology of autism." <em>bioRxiv</em> (2019): 538421. <a href="https://www.biorxiv.org/content/10.1101/484113v3" class="uri">https://www.biorxiv.org/content/10.1101/484113v3</a></p></li>
          <li><p>Huang, Qin, et al. "Delivering genes across the blood-brain barrier: LY6A, a novel cellular receptor for AAV-PHP. B capsids." <em>bioRxiv</em> (2019): 538421. <a href="https://www.biorxiv.org/content/10.1101/538421v1" class="uri">https://www.biorxiv.org/content/10.1101/538421v1</a></p></li>
          <li><p>Kurki, Mitja I., et al. "Contribution of rare and common variants to intellectual disability in a sub-isolate of Northern Finland." <em>Nature Communications</em> 10.1 (2019): 410. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/pmid/30679432/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/pmid/30679432/</a></p></li>
          <li><p>Martin, Alicia R., et al. "Current clinical use of polygenic scores will risk exacerbating health disparities." <em>bioRxiv</em> (2019): 441261. <a href="https://www.biorxiv.org/content/10.1101/441261v3" class="uri">https://www.biorxiv.org/content/10.1101/441261v3</a></p></li>
          <li><p>Collaborative, Epi25, et al. "Ultra-rare genetic variation in the epilepsies: a whole-exome sequencing study of 17,606 individuals." <em>American Journal of Human Genetics</em> (2019): <a href="https://www.cell.com/ajhg/fulltext/S0002-9297(19)30207-1" class="uri">https://www.cell.com/ajhg/fulltext/S0002-9297(19)30207-1</a></p></li>
          <li><p>Karczewski, Konrad J., et al. "The mutational constraint spectrum quantified from variation in 141,456 humans." <em>bioRxiv</em> (2019): 531210. <a href="https://www.biorxiv.org/content/10.1101/531210v4" class="uri">https://www.biorxiv.org/content/10.1101/531210v4</a></p></li>
          <li><p>Whiffin, Nicola, et al. "Human loss-of-function variants suggest that partial LRRK2 inhibition is a safe therapeutic strategy for Parkinsons disease." <em>bioRxiv</em>() (2019): 561472. <a href="https://www.biorxiv.org/content/10.1101/561472v1" class="uri">https://www.biorxiv.org/content/10.1101/561472v1</a></p></li>
          <li><p>Cummings, Beryl B., et al. "Transcript expression-aware annotation improves rare variant discovery and interpretation." <em>bioRxiv</em> (2019): 554444. <a href="https://www.biorxiv.org/content/10.1101/554444v1" class="uri">https://www.biorxiv.org/content/10.1101/554444v1</a></p></li>
          <li><p>Wang, Qingbo, et al. "Landscape of multi-nucleotide variants in 125,748 human exomes and 15,708 genomes." <em>bioRxiv</em> (2019): 573378. <a href="https://www.biorxiv.org/content/10.1101/573378v2" class="uri">https://www.biorxiv.org/content/10.1101/573378v2</a></p></li>
          <li><p>Minikel, Eric Vallabh, et al. "Evaluating potential drug targets through human loss-of-function genetic variation." <em>bioRxiv</em> (2019): 530881. <a href="https://www.biorxiv.org/content/10.1101/530881v2" class="uri">https://www.biorxiv.org/content/10.1101/530881v2</a></p></li>
          <li><p>Collins, Ryan L., et al. "An open resource of structural variation for medical and population genetics." <em>bioRxiv</em> (2019): 578674. <a href="https://www.biorxiv.org/content/10.1101/578674v1" class="uri">https://www.biorxiv.org/content/10.1101/578674v1</a></p></li>
          <li><p>Whiffin, Nicola, et al. "Characterising the loss-of-function impact of 5'untranslated region variants in whole genome sequence data from 15,708 individuals." <em>bioRxiv</em> (2019): 543504. <a href="https://www.biorxiv.org/content/10.1101/543504v1" class="uri">https://www.biorxiv.org/content/10.1101/543504v1</a></p></li>
          <li><p>Lacaze, Paul, et al. "The Medical Genome Reference Bank: a whole-genome data resource of 4000 healthy elderly individuals. Rationale and cohort design." <em>European Journal of Human Genetics</em> 27.2 (2019): 308. <a href="https://www.nature.com/articles/s41431-018-0279-z" class="uri">https://www.nature.com/articles/s41431-018-0279-z</a></p></li>
          <li><p>Cirulli, Elizabeth T., et al. "Genome-wide rare variant analysis for thousands of phenotypes in 54,000 exomes." <em>bioRxiv</em> (2019): 692368. <a href="https://www.biorxiv.org/content/10.1101/692368v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/692368v1.abstract</a></p></li>
          <li><p>Kerminen, Sini, et al. "Geographic Variation and Bias in the Polygenic Scores of Complex Diseases and Traits in Finland." <em>American Journal of Human Genetics</em> (2019). <a href="https://www.biorxiv.org/content/10.1101/485441v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/485441v1.abstract</a></p></li>
          <li><p>Jiang, Fan, Kyle Ferriter, and Claris Castillo. "PIVOT: Cost-Aware Scheduling of Data-Intensive Applications in a Cloud-Agnostic System." <a href="https://renci.org/wp-content/uploads/2019/02/Cloud_19.pdf" class="uri">https://renci.org/wp-content/uploads/2019/02/Cloud_19.pdf</a></p></li>
          <li><p>Pividori, Milton, et al. "Shared and distinct genetic risk factors for childhood-onset and adult-onset asthma: genome-wide and transcriptome-wide studies." <em>Lancet Respiratory Medicine</em> 7.6 (2019): 509-522. <a href="https://www.biorxiv.org/content/10.1101/427427v2" class="uri">https://www.biorxiv.org/content/10.1101/427427v2</a></p></li>
          <li><p>Cox, Samantha L., et al. "Genetic contributions to variation in human stature in prehistoric Europe." <em>bioRxiv</em> (2019): 690545. <a href="https://www.biorxiv.org/content/10.1101/690545v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/690545v1.abstract</a></p></li>
          <li><p>Abrar, Faheem. A Modular Parallel Pipeline Architecture for GWAS Applications in a Cluster Environment. Diss. University of Saskatchewan, 2019. <a href="https://harvest.usask.ca/handle/10388/12087" class="uri">https://harvest.usask.ca/handle/10388/12087</a></p></li>
          <li><p>Khera, Amit V., et al. "Whole-genome sequencing to characterize monogenic and polygenic contributions in patients hospitalized with early-onset myocardial infarction." <em>Circulation</em> 139.13 (2019): 1593-1602. <a href="https://www.ahajournals.org/doi/10.1161/CIRCULATIONAHA.118.035658" class="uri">https://www.ahajournals.org/doi/10.1161/CIRCULATIONAHA.118.035658</a></p></li>
        </ul>
        <h2 id="section-2">2018</h2>
        <ul>
          <li><p>An, Joon-Yong, et al. "Genome-wide de novo risk score implicates promoter variation in autism spectrum disorder." <em>Science</em> (2018): 1. <a href="https://science.sciencemag.org/content/362/6420/eaat6576.full" class="uri">https://science.sciencemag.org/content/362/6420/eaat6576.full</a></p></li>
          <li><p>Molnos, Sophie Claudia. Metabolites: implications in type 2 diabetes and the effect of epigenome-wide interaction with genetic variation. Diss. Technische Universität München, 2018. <a href="https://mediatum.ub.tum.de/1372795f" class="uri">https://mediatum.ub.tum.de/1372795f</a></p></li>
          <li><p>Bis, Joshua C., et al. "Whole exome sequencing study identifies novel rare and common Alzheimer’s-associated variants involved in immune response and transcriptional regulation." <em>Molecular Psychiatry</em> (2018): 1. <a href="https://www.nature.com/articles/s41380-018-0112-7" class="uri">https://www.nature.com/articles/s41380-018-0112-7</a></p></li>
          <li><p>Gormley, Padhraig, et al. "Common variant burden contributes to the familial aggregation of migraine in 1,589 families." <em>Neuron</em> 98.4 (2018): 743-753. <a href="https://www.ncbi.nlm.nih.gov/pubmed/30189203" class="uri">https://www.ncbi.nlm.nih.gov/pubmed/30189203</a></p></li>
          <li><p>Rivas, Manuel A., et al. "Insights into the genetic epidemiology of Crohn's and rare diseases in the Ashkenazi Jewish population." <em>PLoS Genetics</em> 14.5 (2018): e1007329. <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1007329" class="uri">https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1007329</a></p></li>
          <li><p>Satterstrom, F. Kyle, et al. "ASD and ADHD have a similar burden of rare protein-truncating variants." <em>bioRxiv</em> (2018): 277707. <a href="https://www.biorxiv.org/content/10.1101/277707v1" class="uri">https://www.biorxiv.org/content/10.1101/277707v1</a></p></li>
          <li><p>Zekavat, Seyedeh M., et al. "Deep coverage whole genome sequences and plasma lipoprotein (a) in individuals of European and African ancestries." <em>Nature Communications</em> 9.1 (2018): 2606. <a href="https://www.nature.com/articles/s41467-018-04668-w" class="uri">https://www.nature.com/articles/s41467-018-04668-w</a></p></li>
          <li><p>Natarajan, Pradeep, et al. "Deep-coverage whole genome sequences and blood lipids among 16,324 individuals." <em>Nature Communications</em> 9.1 (2018): 3391. <a href="https://www.nature.com/articles/s41467-018-04668-w" class="uri">https://www.nature.com/articles/s41467-018-04668-w</a></p></li>
          <li><p>Farhan, Sali MK, et al. "Enrichment of rare protein truncating variants in amyotrophic lateral sclerosis patients." <em>bioRxiv</em> (2018): 307835. <a href="https://www.biorxiv.org/content/10.1101/307835v2" class="uri">https://www.biorxiv.org/content/10.1101/307835v2</a></p></li>
          <li><p>Ganna, Andrea, et al. "Quantifying the impact of rare and ultra-rare coding variation across the phenotypic spectrum." <em>American Journal of Human Genetics</em> 102.6 (2018): 1204-1211. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5992130/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5992130/</a></p></li>
          <li><p>Khera, Amit V., et al. "Genome-wide polygenic scores for common diseases identify individuals with risk equivalent to monogenic mutations." <em>Nature Genetics</em> 50.9 (2018): 1219. <a href="https://www.nature.com/articles/s41588-018-0183-z?_ga=2.263293700.980063710.1543017600-1151073636.1543017600" class="uri">https://www.nature.com/articles/s41588-018-0183-z?_ga=2.263293700.980063710.1543017600-1151073636.1543017600</a></p></li>
          <li><p>Roselli, Carolina, et al. "Multi-ethnic genome-wide association study for atrial fibrillation." <em>Nature Genetics</em> 50.9 (2018): 1225. <a href="https://www.nature.com/articles/s41588-018-0133-9" class="uri">https://www.nature.com/articles/s41588-018-0133-9</a></p></li>
          <li><p>Arachchi, Harindra, et al. "matchbox: An open‐source tool for patient matching via the Matchmaker Exchange." <em>Human Mutation</em> 39.12 (2018): 1827-1834. <a href="https://onlinelibrary.wiley.com/doi/abs/10.1002/humu.23655" class="uri">https://onlinelibrary.wiley.com/doi/abs/10.1002/humu.23655</a></p></li>
          <li><p>Laisk, Triin, et al. "GWAS meta-analysis highlights the hypothalamic-pituitary-gonadal axis (HPG axis) in the genetic regulation of menstrual cycle length." <em>bioRxiv</em> (2018): 333708. <a href="https://www.biorxiv.org/content/10.1101/333708v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/333708v1.abstract</a></p></li>
          <li><p>Rees, Elliott, et al. "Association between schizophrenia and both loss of function and missense mutations in paralog conserved sites of voltage-gated sodium channels." <em>bioRxiv</em> (2018): 246850. <a href="https://www.biorxiv.org/content/10.1101/246850v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/246850v1.abstract</a></p></li>
          <li><p>Haas, Mary E., et al. "Genetic association of albuminuria with cardiometabolic disease and blood pressure." <em>American Journal of Human Genetics</em> 103.4 (2018): 461-473. <a href="https://www.cell.com/ajhg/pdf/S0002-9297(18)30270-2.pdf" class="uri">https://www.cell.com/ajhg/pdf/S0002-9297(18)30270-2.pdf</a></p></li>
          <li><p>Abel, Haley J., et al. "Mapping and characterization of structural variation in 17,795 deeply sequenced human genomes." <em>bioRxiv</em> (2018): 508515. <a href="https://www.biorxiv.org/content/10.1101/508515v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/508515v1.abstract</a></p></li>
          <li><p>Lane, Jacqueline M., et al. "Biological and clinical insights from genetics of insomnia symptoms." <em>bioRxiv</em> (2018): 257956. <a href="https://www.biorxiv.org/content/10.1101/257956v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/257956v1.abstract</a></p></li>
          <li><p>Pividori, Milton, et al. "Shared and distinct genetic risk factors for childhood onset and adult onset asthma." <em>bioRxiv</em> (2018): 427427. <a href="https://www.biorxiv.org/content/10.1101/427427v1.abstract" class="uri">https://www.biorxiv.org/content/10.1101/427427v1.abstract</a></p></li>
        </ul>
        <h2 id="section-3">2017</h2>
        <ul>
          <li>Lessard, Samuel, et al. "Human genetic variation alters CRISPR-Cas9 on-and off-targeting specificity at therapeutically implicated loci." <em>Proceedings of the National Academy of Sciences</em> 114.52 (2017): E11257-E11266. <a href="https://www.pnas.org/content/114/52/E11257.long" class="uri">https://www.pnas.org/content/114/52/E11257.long</a>
          </li>
        </ul>
        <h2 id="section-4">2016</h2>
        <ul>
          <li>Ganna, Andrea, et al. "Ultra-rare disruptive and damaging mutations influence educational attainment in the general population." <em>Nature Neuroscience</em> 19.12 (2016): 1563. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5127781/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5127781/</a>
          </li>
        </ul>
    </div></div>
  </body>
</html>
